### PR TITLE
Add DynamoDB Streams support

### PR DESCRIPTION
**Pull request by smilerlee**
@nalundgaard and I have finished the implementation, more discuss can be found in #189.
I prefer to preserve the history instead of squashing all these commit together, since not all the commits are made by me.
Anyway, if we need to keep a clean code base history, I can make a squashed PR later.

### CODE REVIEW

**Comment by ransomr** [comment]
Thanks. I won't have a chance to review until next week. I'm ok with not squashing this since it has multiple contributors.

**Comment by nalundgaard** [comment]
+1 to this. Thanks @ransomr for your input and advice on this, and thanks @smilerlee for the implementation, it was fun to collaborate on it.

File: src/erlcloud_ddb_streams.erl [hunk 1]
```diff
@@ -0,0 +1,907 @@
...
+               {"x-amz-target", Operation},
+               {"content-type", "application/x-amz-json-1.0"}],
+    Region =
+        case string:tokens(Config#aws_config.ddb_streams_host, ".") of
+            [_, _, Value, _, _] ->
+                Value;
+            _ ->
+                "us-east-1"
+        end,
+    erlcloud_aws:sign_v4(Config, Headers, Body, Region, "dynamodbstreams").
```
Comment (nalundgaard) [inline src/erlcloud_ddb_streams.erl]: Are you sure this is right? I got this:

```sh
6> {ok, [#ddb_streams_stream{stream_arn=StreamArn} | _]} = erlcloud_ddb_streams:list_streams().
** exception error: no match of right hand side value {error,{<<"InvalidSignatureException">>,
                                                              <<"Credential should be scoped to correct service: 'dynamodb'. ">>}}
```

Comment (smilerlee) [inline src/erlcloud_ddb_streams.erl]: Oops, I misread the aws java sdk source code. It turned out the service name came from [`awssdk_config_default.json`](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-core/src/main/resources/awssdk_config_default.json#L185), not [inferred from the class name](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-core/src/main/java/com/amazonaws/AmazonWebServiceClient.java#L681-L709). Sorry for this.

**Comment by ransomr** [comment]
I'm seeing test failures (on Erlang/OTP 17 [erts-6.4]):

erlcloud_ddb_streams_tests:138: input_test (DescribeStream example request)...*failed*
in function jsx_decoder:key/4 (src/jsx_decoder.erl, line 303)
in call from erlcloud_ddb_streams:request_to_return/1 (src/erlcloud_ddb_streams.erl, line 828)
in call from erlcloud_ddb_streams:describe_stream/3 (src/erlcloud_ddb_streams.erl, line 553)
**error:badarg


erlcloud_ddb_streams_tests:145: input_test (DescribeStream example request with a limit of 1 shard and an exclusive start shard ID)...*failed*
in function jsx_decoder:key/4 (src/jsx_decoder.erl, line 303)
in call from erlcloud_ddb_streams:request_to_return/1 (src/erlcloud_ddb_streams.erl, line 828)
in call from erlcloud_ddb_streams:describe_stream/3 (src/erlcloud_ddb_streams.erl, line 553)
**error:badarg


erlcloud_ddb_streams_tests:206: output_test (DescribeStream example response)...*failed*
in function jsx_decoder:key/4 (src/jsx_decoder.erl, line 303)
in call from erlcloud_ddb_streams:request_to_return/1 (src/erlcloud_ddb_streams.erl, line 828)
in call from erlcloud_ddb_streams:describe_stream/3 (src/erlcloud_ddb_streams.erl, line 553)
in call from erlcloud_ddb_streams_tests:'-output_test/2-fun-1-'/3 (test/erlcloud_ddb_streams_tests.erl, line 115)
**error:badarg

**Comment by nalundgaard** [comment]
@smilerlee said the same thing, but I think it is due to an old jsx. Try removing your deps folder and pulling it down again?

```sh
$ make clean; rm -rf deps
==> erlcloud (clean)
$ . ~/bin/kerl.d/17.5/activate
$ make all
    ...
$ rebar eunit suites=erlcloud_ddb_streams
==> erlcloud (eunit)
Compiled src/erlcloud_xml.erl
src/erlcloud_util.erl:38: Warning: crypto:md5/1 is deprecated and will be removed in in a future release; use crypto:hash/2
Compiled src/erlcloud_util.erl
Compiled src/erlcloud_sts.erl
Compiled src/erlcloud_sqs.erl
Compiled src/erlcloud_sns.erl
Compiled src/erlcloud_ses.erl
Compiled src/erlcloud_retry.erl
src/erlcloud_sdb.erl:33: Warning: export_all flag enabled - all functions will be exported
Compiled src/erlcloud_sdb.erl
Compiled src/erlcloud_mon.erl
Compiled src/erlcloud_kinesis_impl.erl
Compiled src/erlcloud_kinesis.erl
Compiled src/erlcloud_s3.erl
Compiled src/erlcloud_httpc.erl
Compiled src/erlcloud_http.erl
Compiled src/erlcloud_elb.erl
Compiled src/erlcloud_iam.erl
Compiled src/erlcloud_mturk.erl
Compiled src/erlcloud_ddb_util.erl
Compiled src/erlcloud_ddb_impl.erl
Compiled src/erlcloud_ddb_streams.erl
Compiled src/erlcloud_ddb1.erl
Compiled src/erlcloud_ddb.erl
Compiled src/erlcloud_ddb2.erl
Compiled src/erlcloud_cloudtrail.erl
Compiled src/erlcloud_as.erl
Compiled src/erlcloud.erl
Compiled src/erlcloud_aws.erl
Compiled test/erlcloud_sns_tests.erl
Compiled src/erlcloud_ec2.erl
Compiled test/erlcloud_ses_tests.erl
test/erlcloud_sdb_tests.erl:4: Warning: export_all flag enabled - all functions will be exported
Compiled test/erlcloud_sdb_tests.erl
Compiled test/erlcloud_s3_tests.erl
Compiled test/erlcloud_http_tests.erl
Compiled test/erlcloud_kinesis_tests.erl
test/erlcloud_iam_tests.erl:74: Warning: variable 'Key' exported from 'case' (line 67)
test/erlcloud_iam_tests.erl:78: Warning: variable 'Value' exported from 'case' (line 67)
Compiled test/erlcloud_iam_tests.erl
test/erlcloud_ec2_tests.erl:67: Warning: variable 'Key' exported from 'case' (line 60)
test/erlcloud_ec2_tests.erl:71: Warning: variable 'Value' exported from 'case' (line 60)
Compiled test/erlcloud_ec2_tests.erl
Compiled test/erlcloud_ddb_util_tests.erl
Compiled test/erlcloud_ddb_streams_tests.erl
Compiled test/erlcloud_ddb_tests.erl
Compiled test/erlcloud_aws_tests.erl
Compiled test/erlcloud_cloudtrail_tests.erl
Compiled test/erlcloud_as_tests.erl
Compiled test/erlcloud_ddb2_tests.erl
  All 12 tests passed.
```

**Comment by nalundgaard** [comment]
@ransomr I found the comment where @smilerlee discussed it, [here](https://github.com/smilerlee/erlcloud/pull/3#issuecomment-126389839) in smilerlee/erlcloud#3

**Comment by ransomr** [comment]
Thanks - that fixed it. We really should fix the versions of the dependencies.

File: src/erlcloud_ddb_streams.erl [hunk 2]
```diff
@@ -0,0 +1,907 @@
...
+result_fun(#aws_request{response_type = error,
+                        error_type = aws,
+                        response_status = Status,
+                        response_status_line = StatusLine,
+                        response_body = Body} = Request) ->
+    %% == IMPLEMENTATION NOTES ==
+    %% We store the error reason in `httpc_error_reason` for now,
+    %% this may be changed at any time later
+    Request2 = Request#aws_request{
+        httpc_error_reason = {http_error, Status, StatusLine, Body}},
```
Comment (ransomr) [inline src/erlcloud_ddb_streams.erl]: Why is this necessary? Seems like all of this information is stored elsewhere and that this isn't used at all in the implementation. Would prefer to leave it undefined.

**Comment by smilerlee** [comment]
Just got the test cases fixed

**Comment by nalundgaard** [comment]
Wow, that was my bad. I should work on writing better JSON! Thanks for the fix.

File: src/erlcloud_ddb_streams.erl (hunk 2 above)
Comment (smilerlee) [inline src/erlcloud_ddb_streams.erl]: Seems it's a better way, I'll have a detailed think and modify it if necessary a little later.

**Comment by ransomr** [comment]
Thank you for the implementation - great to have this in.
